katsdpcontroller/test/test_product_config.py

```diff
@@ -1,73 +1,1012 @@
+    return BaselineCorrelationProductsStream(
+        'narrow1_bcp', [antenna_channelised_voltage],
+        url=yarl.URL('spead://239.2.0.0+63:7148'),
+        int_time=0.5,
+        n_chans_per_substream=512,
+        n_baselines=40,
+        bits_per_sample=32,
+        instrument_dev_name='narrow1'
+    )
+
+
+def make_tied_array_channelised_voltage(
+        antenna_channelised_voltage: Optional[AntennaChannelisedVoltageStream],
+        name: str, url: yarl.URL) -> TiedArrayChannelisedVoltageStream:
+    if antenna_channelised_voltage is None:
+        antenna_channelised_voltage = make_antenna_channelised_voltage()
+    return TiedArrayChannelisedVoltageStream(
+        name, [antenna_channelised_voltage],
+        url=url,
+        n_chans_per_substream=128,
+        spectra_per_heap=256,
+        bits_per_sample=8,
+        instrument_dev_name='beam'
+    )
+
+
+class TestVisStream:
+    """Test :class:`~.VisStream`."""
+
+    def setup(self) -> None:
+        self.bcp = make_baseline_correlation_products()
+        self.config = {
+            'type': 'sdp.vis',
+            'src_streams': ['narrow1_bcp'],
+            'output_int_time': 1.1,
+            'output_channels': [128, 4096],
+            'continuum_factor': 2,
+            'excise': False,
+            'archive': False
+        }
+
+    def test_from_config(self) -> None:
+        vis = VisStream.from_config(Options(), 'sdp_l0', self.config, [self.bcp], {})
+        assert_equal(vis.int_time, 1.0)   # Rounds to nearest multiple of CBF int_time
+        assert_equal(vis.output_channels, (128, 4096))
+        assert_equal(vis.continuum_factor, 2)
+        assert_equal(vis.excise, False)
+        assert_equal(vis.archive, False)
+        assert_equal(vis.n_servers, 4)
+        assert_is(vis.baseline_correlation_products, self.bcp)
+        assert_equal(vis.n_chans, 1984)
+        assert_equal(vis.n_spectral_chans, 3968)
+        assert_equal(vis.n_spectral_vis, 3968 * 12)
+        assert_equal(vis.antennas, ['m000', 'm002'])
+        assert_equal(vis.n_antennas, 2)
+        assert_equal(vis.n_pols, 2)
+        assert_equal(vis.n_baselines, 12)
+        assert_equal(vis.size, 1984 * (12 * 10 + 4))
+        assert_equal(vis.flag_size, 1984 * 12)
+        assert_equal(vis.data_rate(1.0, 0), vis.size / vis.int_time * 8)
+        assert_equal(vis.flag_data_rate(1.0, 0), vis.flag_size / vis.int_time * 8)
+
+    def test_defaults(self) -> None:
+        del self.config['output_channels']
+        del self.config['excise']
+        vis = VisStream.from_config(Options(), 'sdp_l0', self.config, [self.bcp], {})
+        assert_equal(vis.output_channels, (0, 32768))
+        assert_equal(vis.excise, True)
+
+    def test_bad_continuum_factor(self) -> None:
+        self.config['continuum_factor'] = 3
+        with assert_raises_regex(
+                ValueError,
+                r'n_chans \(32768\) is not a multiple of required alignment \(12\)'):
```

Comment:
(Again with the string testing I know) - how about at least substituting in n_chans and cont_factor * n_servers for the 32768 and 12 values ?

Reply:
Done